yt/frontends/gadget/fields.py

```diff
@@ -40,32 +46,121 @@ def _setup_four_metal_fractions(self, ptype):
         """
         metal_names = ["C", "O", "Si", "Fe"]
 
-        def _Fraction(field, data, i: int):
+        def _fraction(field, data, i: int):
             return data[(ptype, "FourMetalFractions")][:, i]
 
-        def _Metal_density(field, data, i: int):
+        def _metal_density(field, data, i: int):
             return data[(ptype, "FourMetalFractions")][:, i] * data[(ptype, "density")]
 
         for i, metal_name in enumerate(metal_names):
             # add the metal fraction fields
             self.add_field(
                 (ptype, metal_name + "_fraction"),
                 sampling_type="particle",
-                function=partial(_Fraction, i=i),
+                function=partial(_fraction, i=i),
                 units="",
             )
 
             # add the metal density fields
             self.add_field(
                 (ptype, metal_name + "_density"),
                 sampling_type="particle",
-                function=partial(_Metal_density, i=i),
+                function=partial(_metal_density, i=i),
                 units=self.ds.unit_system["density"],
             )
 
         return metal_names
 
-    def _setup_eleven_metal_masses(self, ptype):
+    def _make_fraction_functions(self, ptype, fname):
+        if fname == "ElevenMetalMasses":
+
+            def _fraction(field, data, i: int):
+                return (
+                    data[(ptype, "ElevenMetalMasses")][:, i]
+                    / data[(ptype, "particle_mass")]
+                )
+
+            def _metallicity(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")][:, 1].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return ret
+
+            def _h_fraction(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return 1.0 - ret
+
+            elem_names = ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"]
+
+        elif fname == "MetalMasses":
+            n_elem = len(
+                [
+                    fd
+                    for fd in self.ds.field_list
+                    if fd[0] == ptype and fd[1].startswith("MetalMasses")
+                ]
+            )
+            elem_names = {
+                7: ["C", "N", "O", "Mg", "Si", "Fe", "Ej"],
+                8: ["He", "C", "O", "Mg", "S", "Si", "Fe", "Ej"],
+                11: ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"],
```

Comment:
Out of curiosity, is there a possibility that an output will have both `MetalMasses` and `ElevenMetalMasses` fields?

Reply:
Not really--both are in fact somewhat arbitrary. `ElevenMetalMasses` is a translation field name for Gadget binary datasets defined [here](https://github.com/yt-project/yt/blob/83c96215858a4899e0c0546438c0afca6da12fae/yt/frontends/gadget/definitions.py#L68C3-L85). `MetalMasses` is a convention I've seen used in Gadget HDF5, and (even worse) `Mass of Metals` is one I saw used in another similar dataset.

Reply:
Ok, cool! Thanks for the clarification.